builtins: introduce hlc_to_timestamp

otan opened the pull request:
There is a crdb_internal.approximate_timestamp, which does something
similar but returns a timestamp and is "internal". This is a public
facing version which I think we should keep, returning TimestampTZ so it
returns nicely if the user decides to set the timezone.

Release note (sql change): Introduce a hlc_to_timestamp builtin, which
converts a CockroachDB HLC to a TimestampTZ. This is useful for pretty
printing crdb_internal_mvcc_timestamp or cluster_logical_timestamp(),
but is not useful for accuracy.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/68360)
<!-- Reviewable:end -->

rafiss commented on pkg/sql/sem/tree/eval.go:
```diff
@@ -3695,19 +3695,41 @@ func TimestampToDecimal(ts hlc.Timestamp) apd.Decimal {
+		return nil, err
 	}
-	return TimestampToInexactDTimestamp(hlc.Timestamp{WallTime: coef.Int64()}), nil
+	return TimestampToInexactDTimestamp(ts), nil
```
it seems like we could do away with the `TimestampToInexactDTimestamp` function. so this line would change to
```
return MustMakeDTimestamp(timeutil.Unix(0, ts.WallTime), time.Microsecond), nil
```

(just to make the impls of `DecimalToInexactDTimestampTZ` and `DecimalToInexactDTimestamp` more similar to each other to increase readability.)

otan replied:
`TimestampToInexactDTimestamp` is used in a couple of places in crdb_internal.go, rather not replicate the logic

nvb reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> once Rafi signs off.

> Reviewed 4 of 4 files at r1, 1 of 1 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68360#-:-MgGXWeq-4LLu8rApJUb:bz0tzbb)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @otan and @rafiss)

___
*[pkg/sql/sem/builtins/builtins.go, line 2572 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/68360#-MgGX5I53caJpSy01R8E:-MgGX5I53caJpSy01R8F:b-a0p0oq) ([raw file](https://github.com/cockroachdb/cockroach/blob/5822e4482149e50034cb8364fcb931e8e89b97c6/pkg/sql/sem/builtins/builtins.go#L2572)):*
> ```Go
> 				return tree.DecimalToInexactDTimestampTZ(&d)
> 			},
> 			Info: `Returns a TimestampTZ representation of a CockroachDB HLC in decimal form.
> ```

Should we update the docs on `cluster_logical_timestamp` to reference "CockroachDB HLC in decimal form" as well?


<!-- Sent from Reviewable.io -->

rafiss commented on pkg/sql/sem/tree/casts.go:
```diff
@@ -178,7 +178,7 @@ var validCasts = []castInfo{
 	{
 		from: types.TimestampTZFamily, to: types.StringFamily, volatility: VolatilityStable,
-		volatilityHint: "TIMESTAMPTZ to STRING casts depend on the current timezone; consider using (t AT TIME ZONE 'UTC')::STRING instead.",
+		volatilityHint: "TIMESTAMPTZ to STRING casts depend on the current timezone; consider using (t AT TIME ZONE 'UTC') instead.",
```
looks like this change snuck in, but it's an OK docs fix

otan replied:
i was wondering where this went.

otan committed:
builtins: introduce hlc_to_timestamp

It is worth noting that there is a crdb_internal.approximate_timestamp
builtin, which does something similar but returns a timestamp and is
"internal".

However, this commit introduces a public facing version which I think
we should support and keep, returning TimestampTZ so it
returns nicely if the user decides to set the timezone.

Release note (sql change): Introduce a hlc_to_timestamp builtin, which
converts a CockroachDB HLC to a TimestampTZ. This is useful for pretty
printing crdb_internal_mvcc_timestamp or cluster_logical_timestamp(),
but is not useful for accuracy.

otan commented:
bors r=rafiss,nvanbeschoten
